### PR TITLE
executor: Build via Cloud Build/packer

### LINKED ISSUE #22981: Build VM as much as possible prior to the executor boot script
Try to use Packer to create an image with the correct binaries instead of deferring until VM boot time to build the world.

### PR DESCRIPTION
**Pull request by efritz**
Fixes #22981.

### CODE REVIEW

File: enterprise/cmd/executor/cloudbuild/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+cloudbuild.yaml
```
Comment (eseliger) [inline enterprise/cmd/executor/cloudbuild/.gitignore:1]: not needed, we use a tmp dir

File: enterprise/cmd/executor/build.sh [hunk 2]
```diff
@@ -11,6 +11,9 @@ cleanup() {
 }
 trap cleanup EXIT
 
+# Capture src cli version before we reconfigure go environment.
```
Comment (efritz) [inline enterprise/cmd/executor/build.sh:14]: Why does this need to be before?

File: enterprise/cmd/executor/cloudbuild/build.sh [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+# Ensure image archive path exists
+function ensure_image_archive_path() {
+  mkdir "${EXECUTOR_IMAGE_ARCHIVE_PATH}"
+}
+
+# Write systemd unit file for indexer service
+function install_executor_service() {
+  # Create stub environment file.
+  cat <<EOF >/etc/systemd/system/executor.env
+THIS_ENV_IS="unconfigured"
```
Comment (efritz) [inline enterprise/cmd/executor/cloudbuild/build.sh:160]: Sub environment replaced later by first startup? We should probably put some relevant text/documentation in this file then

File: enterprise/cmd/executor/cloudbuild/build.sh [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+[Unit]
+Description=User code executor
+
+[Service]
+ExecStart=/usr/local/bin/executor
+Restart=always
+EnvironmentFile=/etc/systemd/system/executor.env
+Environment=SRC_LOG_LEVEL=dbug
+Environment=EXECUTOR_IMAGE_ARCHIVE_PATH="${EXECUTOR_IMAGE_ARCHIVE_PATH}" EXECUTOR_FIRECRACKER_IMAGE="${EXECUTOR_FIRECRACKER_IMAGE}"
+Environment=HOME="%h"
```
Comment (efritz) [inline enterprise/cmd/executor/cloudbuild/build.sh:173]: Do multiple envs work fine in this manner?

File: enterprise/cmd/executor/cloudbuild/build.sh [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  systemctl enable executor
+}
+
+###############################
+## THE PLAYBOOK STARTS HERE. ##
+###############################
+install_git
+install_logging_agent
+install_monitoring_agent
+increase_inotify_limit # This must happen before we install docker, because docker might just pick up the limit on startup. (c) efritz 2021.
```
Comment (efritz) [inline enterprise/cmd/executor/cloudbuild/build.sh:189]: will check to see if this is a lie or a truth

File: enterprise/cmd/executor/cloudbuild/build.sh (hunk 4 above)
Comment (eseliger) [reply enterprise/cmd/executor/cloudbuild/build.sh:173]: It did start up, before it wouldn't have. SO told me to do this and I just took it 😄 let's validate before merging

File: enterprise/cmd/executor/cloudbuild/build.sh (hunk 3 above)
Comment (eseliger) [reply enterprise/cmd/executor/cloudbuild/build.sh:160]: yeah I think ideally we disable the entire service in here and then it's not needed. Currently it starts up at VM boot already but isn't configured correctly yet.

File: enterprise/dev/ci/internal/ci/pipeline.go [hunk 6]
```diff
@@ -157,6 +157,7 @@ func GeneratePipeline(c Config) (*bk.Pipeline, error) {
 			triggerAsync(c),               // triggers a slow pipeline, so do it first.
 			addBackendIntegrationTests(c), // ~11m
 			addDockerImages(c, false),     // ~8m (candidate images)
+			addExecutorPackerStep(c),      // very long
```
Comment (mrnugget) [inline enterprise/dev/ci/internal/ci/pipeline.go:160]: Can we only do this if executor actually changed?

File: enterprise/cmd/executor/build.sh [hunk 7]
```diff
@@ -21,9 +24,22 @@ echo "--- go build"
...
+      - 'VERSION=$(git log -n1 --pretty=format:%h)'
+      - 'SRC_CLI_VERSION=$SRC_CLI_VERSION'
+    args: ['build', 'executor.json']
+EOF
+
+# Copy cloudbuild files into workspace.
+cp -R ./cloudbuild/* "$OUTPUT"
+
+# Run gcloud image build.
+gcloud builds submit --config="$OUTPUT/cloudbuild.yaml" "$OUTPUT" --project="sourcegraph-ci"
```
Comment (mrnugget) [inline enterprise/cmd/executor/build.sh:45]: In 1-2 sentences: what are your plans for customers that don't use gcloud? Build different images for different providers?

File: enterprise/dev/ci/internal/ci/pipeline.go (hunk 6 above)
Comment (eseliger) [reply enterprise/dev/ci/internal/ci/pipeline.go:160]: Yes. This PR is WIP, I only threw it in here to see what happens.

File: enterprise/cmd/executor/build.sh (hunk 7 above)
Comment (eseliger) [reply enterprise/cmd/executor/build.sh:45]: There are no set plans yet, this makes it easier and faster for _us_ to deploy it so far, but it still needs the terraform and startup script counterpart from the infrastructure repo. This gets us a bit closer to being able to give it to a customer using Gcloud who's willing to try it out, but it's not the end-state. 

If we want this to be available for customers we should have more than just a Gcloud Option:
- distribute just the binary
- distribute the binary in a Docker image (not the current one, that one has nothing configured and was just used to copy the binary to our runners)
- probably distribute this image and one for AWS, packer makes this relatively straightforward 

And provide extensive documentation. 

But a good step towards CD of our own runners 😊

Comment (mrnugget) [reply enterprise/cmd/executor/build.sh:45]: Thanks! Good stuff. I liked to your commetn from https://docs.google.com/document/d/1BnGF6wQetGZlOb3elvJYRIg44EjmtRQiUu5dRG9q-Nk/edit

File: enterprise/dev/ci/internal/ci/pipeline.go (hunk 6 above)
Comment (efritz) [reply enterprise/dev/ci/internal/ci/pipeline.go:160]: Can we do one of the following for this PR?:

- only do this on main
- do this as a separate async job
- do this based on last tag/git diff (build on change, may be odd if main doesn't build all artfiacts)

**Comment by eseliger** [comment]
TODO: (maybe for later)

We now don't need to pull the big ignite image anymore in the first job run, but we still load two more images, maybe we can do that as part of this here, too:

```
time="2021-07-30T13:37:12Z" level=info msg="docker image \"weaveworks/ignite-kernel:5.10.51\" not found locally, pulling..."
time="2021-07-30T13:37:31Z" level=info msg="Imported OCI image \"weaveworks/ignite-kernel:5.10.51\" (69.6 MB) to kernel image with UID \"786e88f58d384bbd\""
time="2021-07-30T13:37:58Z" level=info msg="Created VM with ID \"df13d5543faf2c93\" and name \"executor-2036b39e-353a-4fd2-a478-f3e024c6ee03\""
time="2021-07-30T13:37:59Z" level=info msg="Pulling image \"weaveworks/ignite:v0.10.0\"..."
time="2021-07-30T13:38:02Z" level=info msg="Networking is handled by \"docker-bridge\""
time="2021-07-30T13:38:02Z" level=info msg="Started Firecracker VM \"df13d5543faf2c93\" in a container with ID \"1b2c8ee0f3d41bafdc8ed2c0e57234f900aca6e011b1d5f79d12a271006f1378\""
time="2021-07-30T13:38:02Z" level=info msg="Waiting for the ssh daemon within the VM to start..."
```

File: enterprise/cmd/executor/cloudbuild/build.sh (hunk 5 above)
Comment (efritz) [reply enterprise/cmd/executor/cloudbuild/build.sh:186]: Eh, cursory search didn't turn anything up, but it doesn't hurt to put something like this _first_ before any installation.

File: enterprise/cmd/executor/build.sh [hunk 8]
```diff
@@ -21,9 +24,22 @@ echo "--- go build"
...
-  --build-arg COMMIT_SHA \
-  --build-arg DATE \
-  --build-arg VERSION
+echo "--- packer build"
+
+cat <<EOF >"$OUTPUT/cloudbuild.yaml"
+steps:
+  - name: gcr.io/cloud-builders/gcloud
+    entrypoint: bash
+    args: ['-c', 'gcloud secrets versions access latest --secret=e2e-builder-sa-key --quiet --project=sourcegraph-ci > /workspace/builder-sa-key.json']
```
Comment (eseliger) [inline enterprise/cmd/executor/build.sh:33]: Should we replace this with a different sa?

File: enterprise/cmd/executor/build.sh (hunk 7 above)
Comment (eseliger) [inline enterprise/cmd/executor/build.sh:45]: Is this the right project to use?

File: enterprise/cmd/executor/cloudbuild/build.sh [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+EOF
+}
+
+###############################
+## THE PLAYBOOK STARTS HERE. ##
+###############################
+install_git
+install_logging_agent
+install_monitoring_agent
+increase_inotify_limit # This must happen before we install docker, because docker might just pick up the limit on startup. (c) efritz 2021.
```
Comment (eseliger) [inline enterprise/cmd/executor/cloudbuild/build.sh:186]:
```suggestion
increase_inotify_limit
```

File: enterprise/dev/ci/internal/ci/pipeline.go (hunk 6 above)
Comment (eseliger) [reply enterprise/dev/ci/internal/ci/pipeline.go:160]: TODO: Only do this on main and perhaps have a "staging" stage, like we do for docker?

Comment (efritz) [reply enterprise/dev/ci/internal/ci/pipeline.go:160]: Where staging would be "build it but don't publish"?

File: enterprise/cmd/executor/build.sh (hunk 8 above)
Comment (efritz) [reply enterprise/cmd/executor/build.sh:33]: cc @sourcegraph/distribution for thoughts

File: enterprise/cmd/executor/build.sh (hunk 7 above)
Comment (efritz) [reply enterprise/cmd/executor/build.sh:45]: Well, it _is_ building _in our CI_, but not _for_ our CI. Not sure if we should just put it in the sourcegraph project or not.

cc @sourcegraph/distribution for thoughts

File: enterprise/dev/ci/internal/ci/pipeline.go (hunk 6 above)
Comment (eseliger) [reply enterprise/dev/ci/internal/ci/pipeline.go:160]: Yeah maybe tag it as `staging-executor-..` or so and do a rename if possible? That way we would not publish a new image if other docker steps in that buildkite waitgroup fail.

File: enterprise/cmd/executor/build.sh (hunk 8 above)
Comment (eseliger) [reply enterprise/cmd/executor/build.sh:33]: Also, CI currently fails because the SA in buildkite doesn't allow to submit a cloud build job.

Comment (efritz) [reply enterprise/cmd/executor/build.sh:33]: Ah yeah, so either way we'll have to add a permissions in the buildkite-agent container as well.

File: enterprise/cmd/executor/build.sh (hunk 7 above)
Comment (davejrt) [reply enterprise/cmd/executor/build.sh:45]: This makes sense. This project has the appropriate tooling/security/APIs to let cloudbuild run so I'd say throw it in there. 

ideally at some point as part of a devex goal all of these builds (including our own) would be run as CI jobs.

**Review by davejrt — APPROVED** [review]

File: enterprise/dev/ci/internal/ci/pipeline.go (hunk 6 above)
Comment (bobheadxi) [reply enterprise/dev/ci/internal/ci/pipeline.go:160]: FYI I think we refer to these as `candidate` images - for Docker, the pipeline [builds candidate images](https://github.com/sourcegraph/sourcegraph/blob/8654ddb423622369ef7b6ea68b1fd2c7850c2cf4/enterprise/dev/ci/internal/ci/pipeline-steps.go#L387:6) that are also [used in tests](https://k8s.sgdev.org/github.com/sourcegraph/sourcegraph@main/-/blob/enterprise/dev/ci/internal/ci/pipeline-steps.go?L291), before retagging those same images and publishing them at the end (https://github.com/sourcegraph/sourcegraph/pull/23364/files#diff-f25e7ee7447fb64c15ac6cb7c50e32e24497f60b316698ae2ff0448551d3f07eR174)

Comment (efritz) [reply enterprise/dev/ci/internal/ci/pipeline.go:160]: Can we "re-tag" the artifacts built by cloud build or would we have to build it a second time under a different name?

As I understand it, the goal here is to have a cloud build item that's accessible for testing/qa pipelines _if you know the name_, but it won't automatically trigger CD or be _obviously_ available for use by enterprise customers. Simply doing a `cp candidate real-image` is what the publish would do, but both would be technically accessible (unless we have stricter org-only perms on candidates) with different understandings of build output quality.

Comment (efritz) [reply enterprise/dev/ci/internal/ci/pipeline.go]: Done.

**Pull request merged**